zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -545,30 +545,29 @@ where
     for payment in proposal.transaction_request().payments() {
         match &payment.recipient_address {
             RecipientAddress::Unified(ua) => {
+                let memo = payment
+                    .memo
+                    .as_ref()
+                    .map_or_else(MemoBytes::empty, |m| m.clone());
                 builder.add_sapling_output(
                     external_ovk,
                     *ua.sapling().expect("TODO: Add Orchard support to builder"),
                     payment.amount,
-                    payment.memo.clone().unwrap_or_else(MemoBytes::empty),
+                    memo.clone(),
                 )?;
                 sapling_output_meta.push((
                     Recipient::Unified(ua.clone(), PoolType::Shielded(ShieldedProtocol::Sapling)),
                     payment.amount,
-                    payment.memo.clone(),
+                    Some(memo),
```

Comment:
Bug: we are now mapping "memo is unset" to "empty memo", and then setting that on a metadata field that has a "memo is unset" state. We should either not do this (preserving the existing behaviour where we internally map the unset state to empty for the builder, but expose the unset state externally), or we should make the metadata field non-optional (as it no longer indicates whether the user actually set a memo or not).

Reply:
This is not a bug; it's an intended change in behavior. If the payment request does not specify a memo, the transaction that is created will have the empty memo, and that is what the recipient will receive. At the payment request level, we can't make the memo required because a `Payment` may be to a transparent address. The goal here is for the notes table(s) to store the same information as actually appears in the transaction.

Reply:
I didn't say this was a bug in the table, but a bug in the metadata. The `Some(memo)` that I left this comment on, ends up being passed to this constructor further down:
https://github.com/zcash/librustzcash/blob/110a9b31d3aebc832bb82537be05047059ee0529/zcash_client_backend/src/data_api/wallet.rs#L635
and is then exposed via this method:
https://github.com/zcash/librustzcash/blob/eade2acab9dcef282ca3298bb14fdcef787bb6e0/zcash_client_backend/src/data_api.rs#L454-L457

Note "if any". That is where the bug lies: the change in this PR is explicitly mapping away the `None` case that would constitute "not any", but the metadata API is still representing to the user that it can and should be expected to occur. The previous code preserved the "if any" case from the part of the code being changed by this PR, to allow the caller to distinguish "implicit empty memo" from "explicit empty memo"; this PR prevents them from doing so. If that is an intentional change, then per my original comment, we should make `SentTransactionOutput::memo` return `&MemoBytes`.

Reply:
We can't make `SentTransactionOutput::memo` return `&MemoBytes` because `SentTransactionOutput` is also used to represent transparent outputs, which cannot have memos.

Reply:
Aha, okay. Then this is a documentation bug, and we should fix the documentation of that method to say that `None` now only means "a transparent output".

Reply:
Fixed in https://github.com/zcash/librustzcash/pull/886/commits/90412ebdc230cbc113945a42ec86124116d09dca